File: test/scalability_1/runtest.pl
#!/usr/bin/perl
#  file     : $HeadURL$
#  revision : $LastChangedRevision$  $LastChangedBy$
#  date     : $LastChangedDate$
#  email    : jdetaeye@users.sourceforge.net

use strict;
use warnings;

use Time::HiRes qw ( gettimeofday tv_interval );
use Env qw(EXECUTABLE);

my %runtimes;

sub createdata (*;$;$;$;$;$;@)
{
  # Pick up the parameters
  my ($OUTFILE,$duplicates,$header,$body,$footer,@subst) = @_;
  my ($curbody, $new, $cnt);

  # Print the header
  printf $OUTFILE "$header";

  # Iteration
  for ($cnt=1; $cnt <= $duplicates; $cnt++)
    {
    $_ = $body;
    for my $cursubst (@subst)
      {
      $new = $cursubst . "_" . $cnt;
      s/$cursubst/$new/g;
      }
    printf $OUTFILE "$_";
    }

  # Finalize
  printf $OUTFILE "$footer";
};


for (my $counter=5000; $counter < 30000; $counter+=5000)
  {
  print "\ncounter $counter\n";
  ## Open a pipe to which all output is sent.
  ## This is an alternative to the flat file proces... The performance of
  ## the program isn't any different.
  # open(OUTFILE, '| time -p $EXECUTABLE');
  open(OUTFILE, ">input.xml");

  my $fh = \*OUTFILE;
  createdata(
    $fh,
    $counter,
    "<PLAN xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\">\n" .
      "<CURRENT>2006-01-01T00:00:00</CURRENT>\n" .
      "<ITEMS>\n",
    "<ITEM NAME=\"ITEMNM\" CATEGORY=\"cat1\" DESCRIPTION=\"DCRP\" >" .
      "\n\t<OPERATION NAME=\"Delivery ITEMNM\" " .
      "xsi:type=\"OPERATION_FIXED_TIME\" DURATION=\"0\"/>" .
      "\n</ITEM>\n",
    "</ITEMS>\n",
    "ITEMNM","DCRP"
    );
  createdata(
    $fh,
    $counter,
    "<OPERATIONS>\n",
    "<OPERATION NAME=\"Make ITEMNM\" xsi:type=\"OPERATION_FIXED_TIME\" " .
      "DURATION=\"24:00:00\"/>\n",
    "</OPERATIONS>\n",
    "ITEMNM",
    );
  createdata(
    $fh,
    $counter,
    "<RESOURCES>\n",
    "<RESOURCE NAME=\"RESNM\"><LOADS>" .
      "<LOAD><OPERATION NAME=\"Make ITEMNM\"/></LOAD></LOADS></RESOURCE>\n",
    "</RESOURCES>\n",
    "RESNM","ITEMNM"
    );
  createdata(
    $fh,
    $counter,
    "<FLOWS>\n",
    "<FLOW xsi:type=\"FLOW_START\"><OPERATION NAME=\"Delivery ITEMNM\"/>" .
      "<BUFFER NAME=\"BUFNM\" ONHAND=\"10\"/>" .
      "<QUANTITY>-1</QUANTITY></FLOW>\n" .
    "<FLOW xsi:type=\"FLOW_END\"><OPERATION NAME=\"Make ITEMNM\"/>" .
      "<BUFFER NAME=\"BUFNM\"/><QUANTITY>1</QUANTITY></FLOW>\n",
    "</FLOWS>\n",
    "ITEMNM","BUFNM"
    );
  createdata(
    $fh,
    $counter,
    "<DEMANDS>\n",
    "<DEMAND NAME=\"DEMANDNM1\" QUANTITY=\"10\" DUE=\"2006-03-03T00:00:00\" " .
     "PRIORITY=\"1\"> <ITEM NAME=\"ITEMNM\"/></DEMAND>\n" .
     "<DEMAND NAME=\"DEMANDNM2\" QUANTITY=\"10\" DUE=\"2006-03-03T00:00:00\" " .
     "PRIORITY=\"2\"> <ITEM NAME=\"ITEMNM\"/></DEMAND>\n" .
     "<DEMAND NAME=\"DEMANDNM3\" QUANTITY=\"10\" DUE=\"2006-03-03T00:00:00\" " .
     "PRIORITY=\"1\"> <ITEM NAME=\"ITEMNM\"/></DEMAND>\n",
    "</DEMANDS></PLAN>\n",
    "DEMANDNM1","DEMANDNM2","DEMANDNM3","ITEMNM"
    );

  close(OUTFILE);

  # Run the execution
  my $starttime = [gettimeofday];
  system("$EXECUTABLE ./commands.xml");
  die "Planner exited abnormally\n" if $? > 0;

  # Measure the time
  $runtimes{$counter} = tv_interval($starttime);
  printf "time: %.3f\n", $runtimes{$counter};

  # Clean up the input and the output
  unlink "input.xml";
  unlink "output.xml";
  #rename "input.xml", "input_$counter.xml";
  #rename "output.xml", "output_$counter.xml";
  }

# Define failure criterium
if ($runtimes{25000} > ($runtimes{5000}+1)*5*1.05)
{
  die "\nTest failed. Run time scales worse than linear with model size.\n";
}

print "\nTest passed\n";
exit;

File: test/scalability_1/runtest.py
#!/usr/bin/env python3
#
# Copyright (C) 2007 by frePPLe bv
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
#
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#
import os
import sys
import subprocess

runtimes = {}


def createdata(outfile, duplicates, header, body, footer, subst):
    # Print the header
    outfile.write(header)

    # Iteration
    if subst == 0:
        for cnt in range(duplicates):
            print(body, file=outfile)
    elif subst == 1:
        for cnt in range(duplicates):
            print(body % (cnt), file=outfile)
    elif subst == 2:
        for cnt in range(duplicates):
            print(body % (cnt, cnt), file=outfile)
    elif subst == 3:
        for cnt in range(duplicates):
            print(body % (cnt, cnt, cnt), file=outfile)
    elif subst == 4:
        for cnt in range(duplicates):
            print(body % (cnt, cnt, cnt, cnt), file=outfile)
    elif subst == 5:
        for cnt in range(duplicates):
            print(body % (cnt, cnt, cnt, cnt, cnt), file=outfile)
    elif subst == 6:
        for cnt in range(duplicates):
            print(body % (cnt, cnt, cnt, cnt, cnt, cnt), file=outfile)
    elif subst == 7:
        for cnt in range(duplicates):
            print(body % (cnt, cnt, cnt, cnt, cnt, cnt, cnt), file=outfile)

    # Finalize
    outfile.write(footer)


# Main loop
for counter in [5000, 10000, 15000, 20000, 25000]:
    print("\ncounter", counter)
    outfile = open("input.xml", "wt")

    createdata(
        outfile,
        counter,
        '<plan xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">\n'
        + "<current>2009-01-01T00:00:00</current>\n"
        + "<items>\n",
        '<item name="ITEMNM_%d" category="cat1" description="DCRP_%d"/>',
        "</items>\n",
        2,
    )
    createdata(
        outfile,
        counter,
        "<operations>\n",
        '<operation name="Make ITEMNM_%d" xsi:type="operation_fixed_time" '
        + 'duration="P1D"><location name="factory"/></operation>',
        "</operations>\n",
        1,
    )
    createdata(
        outfile,
        counter,
        "<resources>\n",
        '<resource name="RESNM_%d"><location name="factory"/><loads>'
        + '<load><operation name="Make ITEMNM_%d"/></load></loads></resource>',
        "</resources>\n",
        2,
    )
    createdata(
        outfile,
        counter,
        "<flows>\n",
        '<flow xsi:type="flow_start"><operation name="Delivery ITEMNM_%d"/>'
        + '<buffer name="BUFNM_%d" onhand="10"/>'
        + "<quantity>-1</quantity></flow>\n"
        + '<flow xsi:type="flow_end"><operation name="Make ITEMNM_%d"/>'
        + '<buffer name="BUFNM_%d"/><quantity>1</quantity></flow>',
        "</flows>\n",
        4,
    )
    createdata(
        outfile,
        counter,
        "<demands>\n",
        '<demand name="DEMANDNM1_%d" quantity="10" due="2009-03-03T00:00:00" '
        + 'priority="1"><location name="factory"/><item name="ITEMNM_%d"/>'
        + '<operation name="Delivery ITEMNM_%d" xsi:type="operation_fixed_time" duration="P0D"/>'
        + "</demand>\n"
        + '<demand name="DEMANDNM2_%d" quantity="10" due="2009-03-03T00:00:00" '
        + 'priority="2"><location name="factory"/><item name="ITEMNM_%d"/></demand>'
        + '<operation name="Delivery ITEMNM_%d" xsi:type="operation_fixed_time" duration="P0D"/>'
        + "</demand>\n"
        + '<demand name="DEMANDNM3_%d" quantity="10" due="2009-03-03T00:00:00" '
        + 'priority="1"><location name="factory"/><item name="ITEMNM_%d"/>'
        + '<operation name="Delivery ITEMNM_%d" xsi:type="operation_fixed_time" duration="P0D"/>'
        + "</demand>\n",
        "</demands></plan>\n",
        9,
    )

    outfile.close()

    # Run the execution
    starttime = os.times()
    with open(os.devnull, "w") as f:
        subprocess.call([os.environ["EXECUTABLE"], "./commands.xml"], stdout=f)

    # Measure the time
    endtime = os.times()
    runtimes[counter] = endtime[4] - starttime[4]
    print("time: %.3f" % (endtime[4] - starttime[4]))

    # Clean up the input and the output
    os.remove("input.xml")
    os.remove("output.xml")
    # if os.path.isfile("input_%d.xml" % counter):
    #  os.remove("input_%d.xml" % counter)
    # os.rename("input.xml", "input_%d.xml" % counter)
    # if os.path.isfile("output_%d.xml" % counter):
    #  os.remove("output_%d.xml" % counter)
    # os.rename("output.xml", "output_%d.xml" % counter)

# Define failure criterium
if runtimes[25000] > runtimes[5000] * 5 * 1.05:
    print("\nTest failed. Run time scales worse than linear with model size.\n")
    sys.exit(1)

print("\nTest passed\n")
